Iter 02 oose backend

QinyuanWu opened the pull request:
1. removed addReview route
2. replaced addReview with request & confirm route
3. plan schema reviewers field changed
4. get plan by plan_id route now return fully populated plan document

MattLiu-mygit reviewed (approved):
This looks good! With this, we can eliminate a lot of the complicated logic in the frontend!

robertzhidealx reviewed (approved):
Looks good! Definitely makes sense to return the full plan directly from the backend.

alanz132 reviewed (approved):
Agreed with Robert, this should make it a lot easier for the frontend.

JJamesWWang commented on model/Plan.js:
```diff
@@ -11,7 +11,13 @@ const planSchema = new Schema({
+  reviewers: [
+    {
+      user_id: { type: String, ref: "User" },
+      pending: { type: Boolean },
```
I think this should be changed to status, which is an enum of PENDING, ACCPTED, APPROVED, etc. This pending var alone does not cover enough state

MattLiu-mygit replied:
This has been noted and will be taken a look at in a later iteration.

JJamesWWang commented on routes/helperMethods.js:
```diff
@@ -9,6 +9,9 @@ function returnData(data, res) {
 //set status code of the response and send error info to the user in json
 function errorHandler(res, status, err) {
+  // console.log("TESTESTTEST");
+  // console.log(res);
```
Remove console log statements?

JJamesWWang commented on routes/majors.spec.ts:
```diff
@@ -1,70 +1,70 @@
-const { returnData } = require("./helperMethods");
-const majors = require("../model/Major");
-const createApp = require("../app");
+// const mongoose = require("mongoose");
```
What's going on here

JJamesWWang commented on routes/planReview.js:
```diff
@@ -15,27 +15,71 @@ router.post("/api/planReview/addReviewer", (req, res) => {
-        //add plan_id to reviewer's doc
-        users
+        //pull up reviewer info
+        await users
```
Do you need `await` if you use `then()`? just a question since I'm not familiar with JS

MattLiu-mygit replied:
No, in fact you don't.

JJamesWWang commented on tests/routes/reviewer.test.js:
```diff
@@ -0,0 +1,46 @@
+
+    // test("Return 200 if successfully added reviewer", async() => {
+
+    // });
```
Remove these?

JJamesWWang reviewed (changes requested):
^See comments above

QinyuanWu commented:
Restructured plan review routes & schemas. removed all review related array fields from User and Plan schema. created new schema PlanReview to keep a separate collection of reviews. please review the pull request again and refer to the updated API documentation. Also, the plan route is no longer populating the whitelisted_plan_ids and reviewers array since they are now removed

JJamesWWang commented on model/PlanReview.js:
```diff
@@ -0,0 +1,21 @@
+  status: {
+    type: String,
+    required: true,
+    enum: ["PENDING", "ACCEPTED", "REJECTED"],
```
Add "APPROVED"; accepted and approved are two different states

MattLiu-mygit replied:
Has this been resolved? @QinyuanWu @JJamesWWang

JJamesWWang replied:
If we go with what's currently here, advisors will not have an option to approve or disapprove of a plan. Up to you.

MattLiu-mygit replied:
This was resolved in Discord and will be made a task in our next Iteration.

JJamesWWang commented on routes/helperMethods.js:
```diff
@@ -9,6 +9,9 @@ function returnData(data, res) {
 //set status code of the response and send error info to the user in json
 function errorHandler(res, status, err) {
+  // console.log("TESTESTTEST");
+  // console.log(res);
```
Remove console log

MattLiu-mygit replied:
Done.

JJamesWWang commented on tests/routes/reviewer.test.js:
```diff
@@ -0,0 +1,46 @@
+
+    // test("Return 200 if successfully added reviewer", async() => {
+
+    // });
```
Remove commented out tests?

QinyuanWu replied:
this is Lawrence's test, not sure if he still needs them

MattLiu-mygit replied:
I've removed it for now.

JJamesWWang reviewed (changes requested):
Looks good, but I think we need to add one more status; see comment

MattLiu-mygit reviewed (approved):
Looks good!

robertzhidealx reviewed (approved):
LGTM

JJamesWWang reviewed (approved):
Looks good

alanz132 reviewed (approved):
Looks good!
